### PR TITLE
Issue 290 export to SVG

### PR DESCRIPTION
**Pull request by serk12**
I've added the SVG "ball and stick" export button. following #290 issue. 
Some important notes are:

-  Installing the  svg-qt libs (libqt5svg5-dev)
- The `frustrumCulling` function is too conservative and I don't know exactly why

### CODE REVIEW

**Comment by ghutchis** [comment]
We use sign-off on our commits - see:
https://github.com/OpenChemistry/avogadrolibs/pull/588/checks?check_run_id=2664267277

File: avogadro/qtplugins/svg/svg.cpp [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+      }
+      QLineF line(from[0] + x, from[1] + y, to[0] + x, to[1] + y);
+      painter.drawLine(line);
+    }
+  }
+}
+
+void SVG::paintMol(QPainter& painter, const Mol& mol)
+{
+  Vector3ub color_mol = m_molecule->color(mol.id);
```
Comment (ghutchis) [inline avogadro/qtplugins/svg/svg.cpp:207]: See, "mol" is usually taken as a short word for "molecule", but this seems to be:
`Vector3ub atom_color = m_molecule->color(atom.id)`

File: avogadro/qtplugins/svg/svg.cpp [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+void SVG::render()
+{
+  if (defaultChecks())
+    return;
+  QString filename = QFileDialog::getSaveFileName(
+    qobject_cast<QWidget*>(parent()), tr("Save File"), QDir::homePath(),
+    tr("SVG (*.svg)"));
+  QFile file(filename);
```
Comment (ghutchis) [inline avogadro/qtplugins/svg/svg.cpp:240]: I don't know if it's only my Mac (likely) but I can save a filename without an SVG extension(!)
```
if (!filename.endsWith(".svg", Qt::CaseInsensitive))
   filename.append(".svg");
```

File: avogadro/qtplugins/svg/svg.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+public slots:
+  void setMolecule(QtGui::Molecule* mol) override;
+  void setScene(Rendering::Scene* scene) override;
+  void setCamera(Rendering::Camera* camera) override;
+
+private slots:
+  void render();
+
+private:
+  struct Mol
```
Comment (ghutchis) [inline avogadro/qtplugins/svg/svg.h:58]: Maybe call it SVGAtom or just Atom

File: avogadro/qtplugins/svg/svg.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  {
+    Eigen::Vector3f pos_model, pos_view, pos_image;
+    unsigned int r;
```
Comment (ghutchis) [inline avogadro/qtplugins/svg/svg.h:61]: radius?

File: avogadro/qtplugins/svg/svg.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  Rendering::Camera* m_camera;
+
+  QAction* m_action;
+  int m_width, m_height;
+  Eigen::Vector3f m_min, m_max;
+  std::vector<Mol> m_mols;
+  std::vector<Eigen::Vector4f> m_frustrum;
+  std::map<unsigned int, unsigned int> m_idToindex;
+
+  static const float RADIUS, PEN_WIDTH_MOL, PEN_WIDTH_BOND, OFF_SET_PARALEL,
```
Comment (ghutchis) [inline avogadro/qtplugins/svg/svg.h:81]: Do these need to be `static`? Why don't we just let these be class members with defaults, and a setOptions() method?

Even if we don't do anything now, someone might decide to add a little "options" dialog to allow user customization.

**Review by ghutchis — CHANGES_REQUESTED** [review]
Overall, it looks good, although my personal preference is no background color. 😄 

I'm trying to figure out if there's a high-res Mac issue because of the scaling ratio. I'll post a some examples.

**Comment by ghutchis** [comment]
Yeah, it's not a resolution / scaling issue - it seems to just be aggressive culling
[caffeine.svg.gz](https://github.com/OpenChemistry/avogadrolibs/files/6541255/caffeine.svg.gz)
[caffeine2.svg.gz](https://github.com/OpenChemistry/avogadrolibs/files/6541264/caffeine2.svg.gz)

File: avogadro/qtplugins/svg/svg.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+{
+  for (const auto& p : m_frustrum) {
+    if ((p.head<3>()).dot(mol.pos_model) + p.w() + mol.r <= 0)
+      return true;
+  }
+  return false;
+}
+
+// transform view position to 2d image
+Eigen::Vector3f SVG::posToImage(const Vector3f& mol, float r)
```
[the remark this answers is not in the document]
Comment (serk12) [reply avogadro/qtplugins/svg/svg.cpp]: Yes Camera::project do almost the same and better. And it fixed the scaling ratio problem (which I didn't saw sorry).

**Comment by serk12** [comment]
I think we have a OS problem on the `m_scene->backgroundColor()` function. 
In my Ubuntu it returning alpha meanwhile your Mac is returning turquoise, a good choice but not as expected.
[cafe.zip](https://github.com/OpenChemistry/avogadrolibs/files/6549454/cafe.zip)

**Comment by ghutchis** [comment]
> I think we have a OS problem on the `m_scene->backgroundColor()` function.
> In my Ubuntu it returning alpha meanwhile your Mac is returning turquoise, a good choice but not as expected.

No, that's the background color I set. 😄

**Comment by ghutchis** [comment]
Code looks much better - @cryos do you want to take a look too?

**Review by ghutchis — APPROVED** [review]

**Comment by ghutchis** [comment]
To make this clear, this fixes #290

File: avogadro/qtplugins/svg/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+avogadro_plugin(SVG
+  "Project the screen in a SVG image."
+  ExtensionPlugin
+  svg.h
+  SVG
+  "svg.cpp"
+  ""
+)
+
+target_link_libraries(SVG LINK_PRIVATE Qt5::Svg)
```
Comment (cryos) [inline avogadro/qtplugins/svg/CMakeLists.txt:14]: Should combine into a single line

File: avogadro/qtplugins/svg/svg.cpp [hunk 8]
```diff
@@ -0,0 +1,277 @@
+#include "svg.h"
```
Comment (cryos) [inline avogadro/qtplugins/svg/svg.cpp:1]: Missing copyright header, consider using the short one in array.h I think

File: avogadro/qtplugins/svg/svg.h [hunk 9]
```diff
@@ -0,0 +1,102 @@
+/******************************************************************************
```
Comment (cryos) [inline avogadro/qtplugins/svg/svg.h:1]: Consider the short copyright header without name or year.

**Review by cryos — COMMENTED** [review]
I need to head out, but took a quick look, haven't had a chance to build/play with it yet. It looks cool, I pointed out a few minor tweaks that you could make. Look forward to playing with it more next week when I have time in front of a compiler.

File: avogadro/qtplugins/svg/CMakeLists.txt (hunk 7 above)
Comment (ghutchis) [reply avogadro/qtplugins/svg/CMakeLists.txt:14]: I think pretty much *all* the plugins have been using multiple lines here, so let's work on that in a separate change.

**Comment by ghutchis** [comment]
Just wanted to mention that it looks good to go. I wanted to wait for @cryos on this one, and it's a holiday weekend here in the States.

**Comment by ghutchis** [comment]
The Mac build is on me - I'll take a look after this is merged.

**Pull request merged**